### PR TITLE
cudaPackages.cudnn: migrate to redist cuda, fix missing zlib

### PR DESCRIPTION
**Pull request by SomeoneSerge**
###### Description of changes

- [x] Removes cudnn's dependency on (runfile-based) `cudatoolkit`, except for temporarily keeping it in `propagatedBuildInputs`
- [x] Adds a `checkPhase` that compares `Runpath` against `DT_NEEDED` via `ldd`
  - [x] Try and replace it with `autoPatchelfHook`?
- [x] Adds missing zlib dependency to `Runpath`

The mentioned missing dependency leads to runtime errors when evaluating jit-traced models in cuda-enabled pytorch

CC @NixOS/cuda-maintainers 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (samuela) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:55]: I guess the central question in terms of whether we can use autopatchelfhook is whether any of these dependencies do not show up in ldd. Otherwise I imagine that it should work.

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:55]: And actually even if there are dependencies that do not show up in ldd, we can always use `addOpenGLRunpath` together with some custom patchelf'ing at the end. So it should def be possible.

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:55]: All of these things are part of DT_NEEDED, all of these things are found through Runpath.
The only thing that isn't part of Runpath by the time I run ldd is libcuda, which isn't part of NEEDED, but must be dlopen()ed at runtime, I guess.

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:55]: I'm not familiar with DT_NEEDED but `libcuda.so.1` has historically been the Achilles heel of autoPatchelfHook. It's my hope that https://github.com/NixOS/nixpkgs/pull/167397 mitigates this pain somewhat.

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:56]: > DT_NEEDED

ELF allows to announce in advance which shared libraries will need to be resolved and loaded at runtime, you can read it e.g. with `patchelf --print-needed lib/libcudnn_cnn_infer.so`. What I wanted to achieve with `find ... ldd | grep` and what I'm now achieving with autoPatchelf is that at least the dependencies declared in this way must be discoverable through runpath. Of course, libcuda has to be handled separately, we just... keep her in mind :(

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:56]: Accidentally, autoPatchelf removes from Rpath things that are not NEEDED -> the problems you mention

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:56]: ah ok so DT_NEEDED is the section that tools like `ldd` inspect to tell us the dependencies?

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:56]: Yes!

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix]: Using autopatchelfhook now, and in a cleaner fashion

**Comment by SomeoneSerge** [comment]
Since I've removed `propagatedBuildInputs` after all, let's wait for the builds to finish: https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/397

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (FRidh) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:92]: can also use the `cudaPackages.cudaVersion attribute`

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (FRidh) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:5]: Because this package is part of cudaPackages I think it is better to pass in the individual deps instead of the whole set.

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:4]: Indeed. I should maybe remove `passthru.cudaPackages` then - downstream derivations should on the opposite query `cudaPackages`, not `cudnn` directly

Comment (FRidh) [reply pkgs/development/libraries/science/math/cudnn/generic.nix]: That's a good one...

**Comment by SomeoneSerge** [comment]
I've copied `autoAddOpenGLRunpath` approach over from https://github.com/NixOS/nixpkgs/pull/168755 and rebuilt the derivation: all seems fine, `/run/opengl-driver/lib` in place, `ldd` has no complaints.

Also re-runing [CI](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/399).

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (FRidh) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:7]: this needs to have ? null for older cudatoolkit versions since the attribute does not exist then

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (FRidh) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:58]: cudatoolkit is still needed for versions where redist packages are not available

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:58]: Are these versions still being supported once 22.05 is released?

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:58]: @FRidh were you meaning to backport the zlib fix? Or is it that you want to keep the function's interface compatible with hypothetical out-of-tree overrides?

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:58]: I'd much rather add a record in release notes for 22.05 and keep things clean. Those who have out-of-tree overrides would get "called with unexpected arguments" error once and update

**Comment by SomeoneSerge** [comment]
I added a mention of the changed function interface in release notes for `22.05`. The premises are as follows:

- Within the current revision all platforms have the `cudaPackages` attribute present (verified with `nix repl`).
- This PR also discards `cudatoolkit` from `propagatedBuildInputs` which is can cause breakages in very unexpected places.
- We've already made very big changes to nixpkgs cuda infrastracture in the previous PRs that are guaranteed to cause some inconsistencies in out-of-tree projects

With all that, I assume it's better to

- Keep this derivation (and all other cuda-redist derivations) as simple and clean as possible, so that they can serve as an example. The maybe-cudatoolkit version grows rather dense to my taste
- Break the obsolete interface now with the release, so that everyone who possibly cares gets the signal and updates accordingly. Rather than they would experience indirect delayed damage
  - I would actually go as far follow up with a PR that marks `cudatoolkit` and `cudnn` attributes in `all-packages.nix` with a `lib.warn` as deprecated.

Of course, this is just my thinking, I definitely want to verify it against more experienced nixpkgs maintainers

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (FRidh) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:58]: > Are these versions still being supported once 22.05 is released? 

We still package them so I suppose so?

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:58]: > We still package them so I suppose so?

/* prolonged miscommunication vibes */ I repeated the word "versions", but I'm really not sure what you were referring to. Versions of what?

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:58]: Aghh 🤦🏻 you mean  cudatolkit<11

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix]: - `libcublas` now optional
- `cudatoolkit` not optional because using passthru `cc`
- cuda versions are present in cudaPackages_10 too

I've checked that all `cudaPackgaes_*.cudnn` evaluate into an `outPath`

**Comment by SomeoneSerge** [comment]
Ok, autoPatchelf has removed `$ORIGIN` and we can't `dlopen("libcudnn_ops_infer.so")` again

**Comment by SomeoneSerge** [comment]
I don't want to fall back to that bash for-loop again =\

1. `patchelf`  needs `--prepend-rpath` and `--apend-rpath`
2. `autoPatchelf` needs `--always-add-libs` or something, that would be adding the selected rpaths unconditionally

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:80]: I'm just going to build and test pytorch with this.
We can remove this later when autoPatchelf UX allows that

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:80]: The test is smth like

```python
#! /usr/bin/env nix-shell
#! nix-shell -i python -p "(import ./. { config = { allowUnfree = true; cudaSupport = true; }; }).python3.withPackages (ps: with ps; [ pytorch ])"

import torch

a = torch.randn(1, 3, 128, 128, device="cuda")
torch.nn.functional.conv2d(a, torch.randn(1, 3, 2, 2, device="cuda"))
```

**Comment by samuela** [comment]
> autoPatchelf needs --always-add-libs or something, that would be adding the selected rpaths unconditionally

Yes, this is a great idea! I was thinking of something similar this morning. I'll try to look into this at some point. I think it would make a lot of people's lives a lot easier.

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:80]: So, this is not crashing anymore.

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:36]: That can easily be moved to the inputs instead of the `? null` trap.

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:40]:
```suggestion
  pname = "cudatoolkit-${cudaMajorVersion}-cudnn";
  inherit version;
```

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:55]: Isn't that done automatically?

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:55]: libcudnn_ops_infer.so.8 may not be in DT_NEEDED/picked up by autoPatchelfHook.

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:36]: You probably mean moving `libcublas` to `buildInputs`, but the intention here is _not to include `cudatoolkit`_ (which explicitly is not redistributable) for cuda>11 where we have the `redist` packages. Thus "`useRedist`"

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:56]: If you mean `getLib` - I think it ends out extracting the default outputs here, I automatically just repeated the pattern I saw somewhere in nixpkgs.
If the question is about the doings of autoPatchelfHook, then yes, there are two factors at play:

1. `autoPatchelf` needs to know in which paths to look for the libraries it finds missing, so we have tell about `gcc.cc`, etc
2. Samuel is correct, `libcudnn_ops_infer.so.8` is not part of NEEDED, which is why we also patch it below (that is more of a temporary measure though)

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:81]: To be clear: my assumption here was that it's always `libcudnn.so` importing other libraries in `$out/lib`, and not the other way around. Only `libcudnn.so` ends up with `$out/lib` in its runpath. Before the PR all of the libraries knew about `$out/lib`, but that may have been redundant.

The test verifies this assumption in a very simplistic setting, but I'm confident enough that this doesn't break things

File: pkgs/development/libraries/science/math/cudnn/extension.nix
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/cudnn/extension.nix:10]: I'd move it to `args`, but I'd have to split `8.3.2.44` into two instances - pre and post cuda 11.4 - and adjust the selector expression smh... I'll keep it this way, I guess

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix]: I thought it actually is a good idea to move the cudatoolkit-vs-redist flag to the top (I mean, it seems at least one person got confused about the meaning of the previous flag, so I might as well make it more explicit), It now has a scary name "use ... runfile", should give people the idea that we'd like to get rid of this

**Comment by SomeoneSerge** [comment]
@SuperSandro2000 @FRidh would you have any objections now?

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:59]:
```suggestion
  ] ++ lib.optionals useCudatoolkitRunfile [
    libcublas
```

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:8]:
```suggestion
, libcublas # cuda <11 doesn't ship redist packages
```

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:59]: No, the point is exactly that `libcublas` is _the redistributable alternative_ to run-file based `cudatoolkit`

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:8]: This must be `? null` because for `cudaPacakges_{10,...,11_3}` there's no `libcublas`

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:39]: May seem redundant, but the idea is to repeat a pattern used elsewhere anyway, e.g. in pytorch. The pattern being: that packages use cuda as a "toolchain" with chosen hosts and targets, and we decide in the derivation which modules to include in the toolchain. Runfile being a maybe-non-redist throw-all-in toolchain.

Higher-level goal: I want people like Sandro (maintainers but out of narrow cuda context) not to get confused when just skimming through changes

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:41]: [OfBorg doesn't particularly like this](https://gist.github.com/GrahamcOfBorg/bbf9fba8fe9b8821d2a5e6f321ff0b6c)

**Comment by SomeoneSerge** [comment]
@FRidh @SuperSandro2000 I'd like to either get new change requests or a merge, because right now pytorch and cudnn are still broken (runtime errors) on master[^1]

[^1]: ...I'm still trying to distill my failure example into something I can publish, but so far not too successful

**Comment by FRidh** [comment]
warnings are not allowed and need to be removed. Other than that, it looks good to me.

File: pkgs/development/libraries/science/math/cudnn/generic.nix
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cudnn/generic.nix]: Deleted `lib.warn`

**Comment by SomeoneSerge** [comment]
I removed the warnings, rebased and squashed the branch to remove the noise, fixed an error with `cudnn_7_6_5` that I only noticed with `nixpkgs-review`.

It seems the full `nixpkgs-review` is going to take ages (triggered tensorflow, jaxlib, etc).
I think I may have broken `katago` - it may have relied on cudnn's `propagatedBuildInputs`.
I'll see if I can fix it quick, or if I should mark it as conditionally broken

**Comment by SomeoneSerge** [comment]
```
12 packages updated:
caffe caffe caffe katago python3.10-cupy python3.10-pytorch python3.10-Theano python3.9-cupy python3.9-jaxlib python3.9-pytorch python3.9-tensorflow-gpu python3.9-Theano
```

**Comment by SomeoneSerge** [comment]
@FRidh Ready

**Review by samuela — APPROVED** [review]
changes LGTM. @FRidh @SuperSandro2000 are we good to merge?

**Pull request merged**
